library/core/src/iter/traits/iterator.rs

```diff
@@ -2217,6 +2218,61 @@ pub trait Iterator {
         Some(self.fold(first, f))
     }
 
+    /// Reduces the elements to a single one by repeatedly applying a reducing
+    /// operation. If the closure returns a failure, the error is propagated
+    /// back to the caller immediately.
+    ///
+    /// [`None`] is returned when the iterator is empty.
+    ///
+    /// For iterators with at least one element, this is the same as calling
+    /// [`try_fold()`] with the first element of the iterator as the initial
+    /// accumulator value.
+    ///
+    /// [`try_fold()`]: Iterator::try_fold
+    ///
+    /// # Examples
+    ///
+    /// Safely calculate the sum of a series of numbers:
+    ///
+    /// ```
+    /// #![feature(iterator_try_reduce)]
+    ///
+    /// let numbers: Vec<usize> = vec![10, 20, 5, 23, 0];
+    /// let sum = numbers.into_iter().try_reduce(|x, y| x.checked_add(y));
+    /// assert_eq!(sum, Some(Some(58)));
+    /// ```
+    ///
+    /// Determine when a reduction was short circuited:
+    ///
+    /// ```
+    /// #![feature(iterator_try_reduce)]
+    ///
+    /// let numbers = vec![1, 2, 3, usize::MAX, 4, 5];
+    /// let sum = numbers.into_iter().try_reduce(|x, y| x.checked_add(y));
+    /// assert_eq!(sum, Some(None));
+    /// ```
+    ///
+    /// Determine when a reduction was not performed because there are no elements:
+    ///
+    /// ```
+    /// #![feature(iterator_try_reduce)]
+    ///
+    /// let numbers: Vec<usize> = Vec::new();
+    /// let sum = numbers.into_iter().try_reduce(|x, y| x.checked_add(y));
+    /// assert_eq!(sum, None);
+    /// ```
+    #[inline]
+    #[unstable(feature = "iterator_try_reduce", reason = "new API", issue = "87053")]
+    fn try_reduce<F, R>(&mut self, f: F) -> Option<R>
```

Comment:
Pondering: It's not clear to me that getting an `Option<Result<T, E>>` from this is what people would usually want.  Sure, they can do `.try_reduce(...).transpose()?`, but I wonder if there's a different approach that could make it more convenient.

[`try_find`](https://doc.rust-lang.org/nightly/std/iter/trait.Iterator.html#method.try_find) is the other "fallible terminal implemented by wrapping `try_fold`", so maybe `try_reduce` wants the same HKT-like thing that `try_map` does (see #85115).

Reply:
This pull request exists because I wanted the equivalent of `try_fold` or `try_find` but for `reduce` and I couldn't find it. It seemed inconsistent to not have an implementation for `reduce`.

I don't know if the return type for these functions is ideal. I don't know enough about the language or its development to have an informed opinion. If adding `try_reduce` for the sake of consistency isn't an acceptable reason, I would be more than happy to see this pull request closed.

Reply:
I've opened https://github.com/rust-lang/rust/pull/91286 with a way to solve this kind of problem.

With that, instead of `-> Option<R>` you'd have `-> ChangeOutputType<R, Option<R::Output>>`.  That would mean that if the closure returns `Result<T, E>` you'll get `Result<Option<T>, E>` -- and in a way that `try_reduce(...)?` would infer correctly.

Reply:
@kit-981 That PR has now landed, so if you want to update this one to use those facilities, I'd be happy to review it.  Feel free to ping me here or on zulip if you have questions about how to get it to work.

Reply:
Will do. Thank you!

Reply:
Update looks good here 👍